src/lib.rs

```diff
@@ -256,71 +261,89 @@ impl SubnetActor for Actor {
-    //             Method::CommitChildCheckpoint as u64,
-    //             RawBytes::serialize(checkpoint)?,
-    //             0.into(),
-    //         )?;
-    //         // remove votes used for commitment
-    //         if found {
-    //             votes_map.delete(&ch_cid.to_bytes())?;
-    //         }
-    //     } else {
-    //         // if no majority store vote and return
-    //         votes_map.set(ch_cid.to_bytes().into(), votes)?;
-    //     }
-    //
-    //     // flush votes
-    //     st.window_checks = votes_map.flush()?;
-    //
-    //     st.save();
-    //     Ok(None)
-    // }
+    /// SubmitCheckpoint accepts signed checkpoint votes for miners.
+    ///
+    /// This functions verifies that the checkpoint is valid before
+    /// propagating it for commitment to the IPC gateway. It expects at least
+    /// votes from 2/3 of miners with collateral.
+    fn submit_checkpoint<BS, RT>(
+        rt: &mut RT,
+        ch: Checkpoint,
+    ) -> Result<Option<RawBytes>, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let state: State = rt.state()?;
+        let caller = rt.message().caller();
+
+        if !state.is_validator(&caller) {
+            return Err(actor_error!(illegal_state, "not validator"));
+        }
+
+        state
+            .verify_checkpoint(rt, &ch)
+            .map_err(|_| actor_error!(illegal_state, "checkpoint failed"))?;
+
+        let mut msg = None;
+
+        rt.transaction(|st: &mut State, rt| {
+            let ch_cid = ch.cid();
+
+            let mut found = false;
+            let mut votes = match st.get_votes(rt.store(), &ch_cid)? {
+                Some(v) => {
+                    found = true;
+                    v
+                }
+                None => Votes {
+                    validators: Vec::new(),
+                },
+            };
+
+            if votes.validators.iter().any(|x| x == &caller) {
+                return Err(actor_error!(
+                    illegal_state,
+                    "miner has already voted the checkpoint"
+                ));
+            }
+
+            // add miner vote
+            votes.validators.push(caller);
+
+            // if has majority
+            if st.has_majority_vote(rt.store(), &votes)? {
+                // commit checkpoint
+                st.flush_checkpoint(rt.store(), &ch)
+                    .map_err(|_| actor_error!(illegal_state, "cannot flush checkpoint"))?;
+
+                // prepare the message
+                msg = Some(CrossActorPayload::new(
+                    st.ipc_gateway_addr,
+                    ipc_gateway::Method::CommitChildCheckpoint as u64,
+                    RawBytes::serialize(ch)?,
+                    TokenAmount::zero(),
+                ));
+
+                // remove votes used for commitment
+                if found {
+                    st.remove_votes(rt.store(), &ch_cid)?;
```

Comment:
We need to ensure (and I think this is the case) that the commitment and the cleaning of the votes is atomic, if not it may lead to some nasty bugs/potential vulnerability.

Reply:
I thought we have it wrapped in `transaction`?

Reply:
Yes, this is atomic because we actually `verify_checkpoint` before the transaction, so we are good to go.